scopesim/source/source_fields.py

```diff
@@ -0,0 +1,314 @@
+        y += quantify(dy, u.arcsec)
+        self.field["y"] = y
+
+
+@dataclass
+class HDUSourceField(SourceField):
+    """Base class for source fields with HDU."""
+
+    field: fits.ImageHDU
+    wcs: WCS = dataclass_field(default_factory=WCS, init=False)
+
+    def __new__(cls, *args, **kwargs):
+        """Override creation to create subclasses."""
+        if issubclass(cls, (CubeSourceField, ImageSourceField)):
+            # Allow for dirct subclass access
+            return super().__new__(cls)
+
+        field = kwargs.get("field", args[0])
+        if field.header["NAXIS"] == 3:
+            return super().__new__(CubeSourceField)
+        if field.header["NAXIS"] == 2:
+            return super().__new__(ImageSourceField)
+
+        # If we get here, something went wrong
+        raise TypeError(f"{field.header['NAXIS'] = } must be 2 or 3.")
+
+    @property
+    def header(self) -> fits.Header:
+        """Shortcut for `field.header`."""
+        return self.field.header
+
+    @property
+    def data(self) -> np.ndarray:
+        """Shortcut for `field.data`."""
+        return self.field.data
+
+    @data.setter
+    def data(self, value):
+        self.field.data = value
+
+    @property
+    def img_size(self) -> str:
+        """Shortcut for `field.data.shape`."""
+        if self.data is None:
+            return "<empty>"
+        return str(self.data.shape)
+
+    def _write_stream(self, stream: TextIO) -> None:
+        stream.write(f"ImageHDU with size {self.img_size}, referencing "
+                     f"spectrum {self.field.header.get('SPEC_REF', '-')}")
+
+    def plot(self, axes, color) -> None:
+        """Plot source."""
+        xypts = imp_utils.calc_footprint(self.header)
+        convf = u.Unit(self.header["CUNIT1"]).to(u.arcsec)
+        outline = np.array(list(close_loop(xypts))) * convf
+        axes.plot(*outline.T, color, label=self.name)
+
+    def shift(self, dx, dy) -> None:
+        """Shift source by dx, dy."""
+        dx = dx << u.arcsec << self.wcs.wcs.cunit[0]
+        dy = dy << u.arcsec << self.wcs.wcs.cunit[1]
+        self.header["CRVAL1"] += dx.value
+        self.header["CRVAL2"] += dy.value
+
+
+@dataclass
+class ImageSourceField(SpectrumSourceField, HDUSourceField):
+    """Source field with 2D image and a single (average) spectrum."""
+
+    def __post_init__(self):
+        """Validate input."""
+        assert self.spectra, "Spectra must be non-empty for 2D image source."
+        try:
+            self.wcs = WCS(self.field)
+        except (SingularMatrixError, FITSFixedWarning):
+            # This occurs for BG SRC
+            logger.debug("Couldn't create source field WCS.")
+            self.wcs = None
+
+
+@dataclass
+class CubeSourceField(HDUSourceField):
+    """Source field with 3D data cube."""
+
+    def __post_init__(self):
+        """Validate input."""
+        if self.wcs is None and not self.meta.get("from_hdul", False):
+            self.wcs = WCS(self.field)
```

Comment:
I can't really in good faith let you add more to `self.meta` :stuck_out_tongue_winking_eye: .
```suggestion
class CubeSourceField(HDUSourceField):
    """Source field with 3D data cube."""
    from_hdul: bool = False

    def __post_init__(self):
        """Validate input."""
        if self.wcs is None and self.from_hdul:
            self.wcs = WCS(self.field)
```
(Perhaps this needs to be `kw_only`?)